{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# variables to be modified by the user\n",
    "    # input paths\n",
    "VIBRANT_output_path = '/home/abelardoacm/Storage/Abelardo/projects/mini-devel/results/1_biome_fasta_files/biome0/VIBRANT_DTRs_20kb/'\n",
    "project_root = '/home/abelardoacm/Storage/Abelardo/projects/mini-devel/'\n",
    "JGI_global_metadata_csv_path = project_root + 'data/1_environmental_seqs_from_JGI_IMGvir/DTRs_20kb.csv'\n",
    "    # parameters\n",
    "min_quality_by_usr = \"complete\" # 'low', 'medium', 'high', 'complete'\n",
    "min_contig_length_by_usr = 0 # in bp\n",
    "max_contig_length_by_usr = 5000 # in bp\n",
    "mcp_terms_by_usr = None #(list, optional): Terms related to MCP. Defaults to a predefined list if None.\n",
    "false_terms_by_usr = None #(list, optional): Terms to identify and exclude MPC false positives. Defaults to a predefined list if None."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create the output directory if it does not exist\n",
    "output_directory = VIBRANT_output_path + '2_filtered_VIBRANT_output/'\n",
    "if not os.path.exists(output_directory):\n",
    "    os.makedirs(output_directory)\n",
    "os.chdir(project_root)\n",
    "VIBRANT_output_path_elements = VIBRANT_output_path.split('/')\n",
    "VIBRANT_folder_name = next((element for element in reversed(VIBRANT_output_path_elements) if element), None)\n",
    "VIBRANT_input_name = '_'.join(VIBRANT_folder_name.split('_')[1:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load the JGI global metadata\n",
    "JGI_global_metadata = pd.read_csv(JGI_global_metadata_csv_path)\n",
    "\n",
    "# See VIBRANT Output explanations for more information on the output files and numbers\n",
    "# output 38 is the VIBRANT annotation file\n",
    "output38_file_path =  VIBRANT_output_path + 'VIBRANT_results_' + VIBRANT_input_name +'/VIBRANT_annotations_' + VIBRANT_input_name + '.tsv'\n",
    "VIBRANT_annotations = pd.read_csv(output38_file_path, sep='\\t')\n",
    "\n",
    "# output 42 is the list of predicted genome quality and type\n",
    "output42_file_path =  VIBRANT_output_path + 'VIBRANT_results_' + VIBRANT_input_name +'/VIBRANT_genome_quality_' + VIBRANT_input_name + '.tsv'\n",
    "VIBRANT_genome_quality = pd.read_csv(output42_file_path, sep='\\t')\n",
    "\n",
    "# output 23 corresponds to all encoded proteins among identified phages\n",
    "output23_file_path =  VIBRANT_output_path + 'VIBRANT_phages_' + VIBRANT_input_name + '/' + VIBRANT_input_name + '.phages_combined.faa'\n",
    "VIBRANT_phages_proteins = SeqIO.parse(output23_file_path, \"fasta\") # read the sequences from the fasta file\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>scaffold</th>\n",
       "      <th>KO</th>\n",
       "      <th>AMG</th>\n",
       "      <th>KO name</th>\n",
       "      <th>KO evalue</th>\n",
       "      <th>KO score</th>\n",
       "      <th>KO v-score</th>\n",
       "      <th>Pfam</th>\n",
       "      <th>Pfam name</th>\n",
       "      <th>Pfam evalue</th>\n",
       "      <th>Pfam score</th>\n",
       "      <th>Pfam v-score</th>\n",
       "      <th>VOG</th>\n",
       "      <th>VOG name</th>\n",
       "      <th>VOG evalue</th>\n",
       "      <th>VOG score</th>\n",
       "      <th>VOG v-score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>DTR_156067_2</td>\n",
       "      <td>DTR_156067</td>\n",
       "      <td>K06904</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K06904; uncharacterized protein</td>\n",
       "      <td>5.000000e-41</td>\n",
       "      <td>140.3</td>\n",
       "      <td>2.41</td>\n",
       "      <td>PF05065.13</td>\n",
       "      <td>Phage capsid family</td>\n",
       "      <td>7.200000e-39</td>\n",
       "      <td>133.0</td>\n",
       "      <td>5.12</td>\n",
       "      <td>VOG00633</td>\n",
       "      <td>sp|O64210|CAPSD_BPMD2 Probable major capsid pr...</td>\n",
       "      <td>1.500000e-45</td>\n",
       "      <td>155.6</td>\n",
       "      <td>10.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213</th>\n",
       "      <td>DTR_381167_12</td>\n",
       "      <td>DTR_381167</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PF05065.13</td>\n",
       "      <td>Phage capsid family</td>\n",
       "      <td>2.100000e-37</td>\n",
       "      <td>128.2</td>\n",
       "      <td>5.12</td>\n",
       "      <td>VOG00633</td>\n",
       "      <td>sp|O64210|CAPSD_BPMD2 Probable major capsid pr...</td>\n",
       "      <td>1.400000e-37</td>\n",
       "      <td>129.4</td>\n",
       "      <td>10.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>543</th>\n",
       "      <td>DTR_885259_17</td>\n",
       "      <td>DTR_885259</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>VOG02473</td>\n",
       "      <td>sp|G9M952|CAPSD_BPPS4 Major capsid protein</td>\n",
       "      <td>1.400000e-91</td>\n",
       "      <td>306.7</td>\n",
       "      <td>1.19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>580</th>\n",
       "      <td>DTR_316164_22</td>\n",
       "      <td>DTR_316164</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PF05065.13</td>\n",
       "      <td>Phage capsid family</td>\n",
       "      <td>2.300000e-20</td>\n",
       "      <td>72.3</td>\n",
       "      <td>5.12</td>\n",
       "      <td>VOG00633</td>\n",
       "      <td>sp|O64210|CAPSD_BPMD2 Probable major capsid pr...</td>\n",
       "      <td>1.900000e-16</td>\n",
       "      <td>59.7</td>\n",
       "      <td>10.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>618</th>\n",
       "      <td>DTR_166448_17</td>\n",
       "      <td>DTR_166448</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PF05065.13</td>\n",
       "      <td>Phage capsid family</td>\n",
       "      <td>9.800000e-18</td>\n",
       "      <td>63.7</td>\n",
       "      <td>5.12</td>\n",
       "      <td>VOG00633</td>\n",
       "      <td>sp|O64210|CAPSD_BPMD2 Probable major capsid pr...</td>\n",
       "      <td>2.800000e-23</td>\n",
       "      <td>82.2</td>\n",
       "      <td>10.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93452</th>\n",
       "      <td>DTR_282867_5</td>\n",
       "      <td>DTR_282867</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>VOG01505</td>\n",
       "      <td>REFSEQ capsid protein</td>\n",
       "      <td>2.700000e-13</td>\n",
       "      <td>49.7</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93501</th>\n",
       "      <td>DTR_545965_14</td>\n",
       "      <td>DTR_545965</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PF05065.13</td>\n",
       "      <td>Phage capsid family</td>\n",
       "      <td>1.500000e-41</td>\n",
       "      <td>141.9</td>\n",
       "      <td>5.12</td>\n",
       "      <td>VOG00633</td>\n",
       "      <td>sp|O64210|CAPSD_BPMD2 Probable major capsid pr...</td>\n",
       "      <td>2.000000e-45</td>\n",
       "      <td>155.3</td>\n",
       "      <td>10.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93625</th>\n",
       "      <td>DTR_890275_1</td>\n",
       "      <td>DTR_890275</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PF05065.13</td>\n",
       "      <td>Phage capsid family</td>\n",
       "      <td>4.200000e-14</td>\n",
       "      <td>51.8</td>\n",
       "      <td>5.12</td>\n",
       "      <td>VOG00633</td>\n",
       "      <td>sp|O64210|CAPSD_BPMD2 Probable major capsid pr...</td>\n",
       "      <td>6.100000e-20</td>\n",
       "      <td>71.3</td>\n",
       "      <td>10.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94005</th>\n",
       "      <td>DTR_348973_11</td>\n",
       "      <td>DTR_348973</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PF11651.8</td>\n",
       "      <td>P22 coat protein - gene protein 5</td>\n",
       "      <td>2.700000e-12</td>\n",
       "      <td>45.3</td>\n",
       "      <td>1.00</td>\n",
       "      <td>VOG00241</td>\n",
       "      <td>sp|P19727|CAPSB_BPT7 Minor capsid protein</td>\n",
       "      <td>2.500000e-33</td>\n",
       "      <td>115.1</td>\n",
       "      <td>10.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94134</th>\n",
       "      <td>DTR_696999_3</td>\n",
       "      <td>DTR_696999</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PF05065.13</td>\n",
       "      <td>Phage capsid family</td>\n",
       "      <td>1.000000e-41</td>\n",
       "      <td>142.5</td>\n",
       "      <td>5.12</td>\n",
       "      <td>VOG00633</td>\n",
       "      <td>sp|O64210|CAPSD_BPMD2 Probable major capsid pr...</td>\n",
       "      <td>3.400000e-68</td>\n",
       "      <td>230.3</td>\n",
       "      <td>10.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>864 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             protein    scaffold      KO  AMG  \\\n",
       "26      DTR_156067_2  DTR_156067  K06904  NaN   \n",
       "213    DTR_381167_12  DTR_381167     NaN  NaN   \n",
       "543    DTR_885259_17  DTR_885259     NaN  NaN   \n",
       "580    DTR_316164_22  DTR_316164     NaN  NaN   \n",
       "618    DTR_166448_17  DTR_166448     NaN  NaN   \n",
       "...              ...         ...     ...  ...   \n",
       "93452   DTR_282867_5  DTR_282867     NaN  NaN   \n",
       "93501  DTR_545965_14  DTR_545965     NaN  NaN   \n",
       "93625   DTR_890275_1  DTR_890275     NaN  NaN   \n",
       "94005  DTR_348973_11  DTR_348973     NaN  NaN   \n",
       "94134   DTR_696999_3  DTR_696999     NaN  NaN   \n",
       "\n",
       "                               KO name     KO evalue  KO score  KO v-score  \\\n",
       "26     K06904; uncharacterized protein  5.000000e-41     140.3        2.41   \n",
       "213                                NaN           NaN       NaN         NaN   \n",
       "543                                NaN           NaN       NaN         NaN   \n",
       "580                                NaN           NaN       NaN         NaN   \n",
       "618                                NaN           NaN       NaN         NaN   \n",
       "...                                ...           ...       ...         ...   \n",
       "93452                              NaN           NaN       NaN         NaN   \n",
       "93501                              NaN           NaN       NaN         NaN   \n",
       "93625                              NaN           NaN       NaN         NaN   \n",
       "94005                              NaN           NaN       NaN         NaN   \n",
       "94134                              NaN           NaN       NaN         NaN   \n",
       "\n",
       "             Pfam                          Pfam name   Pfam evalue  \\\n",
       "26     PF05065.13                Phage capsid family  7.200000e-39   \n",
       "213    PF05065.13                Phage capsid family  2.100000e-37   \n",
       "543           NaN                                NaN           NaN   \n",
       "580    PF05065.13                Phage capsid family  2.300000e-20   \n",
       "618    PF05065.13                Phage capsid family  9.800000e-18   \n",
       "...           ...                                ...           ...   \n",
       "93452         NaN                                NaN           NaN   \n",
       "93501  PF05065.13                Phage capsid family  1.500000e-41   \n",
       "93625  PF05065.13                Phage capsid family  4.200000e-14   \n",
       "94005   PF11651.8  P22 coat protein - gene protein 5  2.700000e-12   \n",
       "94134  PF05065.13                Phage capsid family  1.000000e-41   \n",
       "\n",
       "       Pfam score  Pfam v-score       VOG  \\\n",
       "26          133.0          5.12  VOG00633   \n",
       "213         128.2          5.12  VOG00633   \n",
       "543           NaN           NaN  VOG02473   \n",
       "580          72.3          5.12  VOG00633   \n",
       "618          63.7          5.12  VOG00633   \n",
       "...           ...           ...       ...   \n",
       "93452         NaN           NaN  VOG01505   \n",
       "93501       141.9          5.12  VOG00633   \n",
       "93625        51.8          5.12  VOG00633   \n",
       "94005        45.3          1.00  VOG00241   \n",
       "94134       142.5          5.12  VOG00633   \n",
       "\n",
       "                                                VOG name    VOG evalue  \\\n",
       "26     sp|O64210|CAPSD_BPMD2 Probable major capsid pr...  1.500000e-45   \n",
       "213    sp|O64210|CAPSD_BPMD2 Probable major capsid pr...  1.400000e-37   \n",
       "543           sp|G9M952|CAPSD_BPPS4 Major capsid protein  1.400000e-91   \n",
       "580    sp|O64210|CAPSD_BPMD2 Probable major capsid pr...  1.900000e-16   \n",
       "618    sp|O64210|CAPSD_BPMD2 Probable major capsid pr...  2.800000e-23   \n",
       "...                                                  ...           ...   \n",
       "93452                              REFSEQ capsid protein  2.700000e-13   \n",
       "93501  sp|O64210|CAPSD_BPMD2 Probable major capsid pr...  2.000000e-45   \n",
       "93625  sp|O64210|CAPSD_BPMD2 Probable major capsid pr...  6.100000e-20   \n",
       "94005          sp|P19727|CAPSB_BPT7 Minor capsid protein  2.500000e-33   \n",
       "94134  sp|O64210|CAPSD_BPMD2 Probable major capsid pr...  3.400000e-68   \n",
       "\n",
       "       VOG score  VOG v-score  \n",
       "26         155.6        10.00  \n",
       "213        129.4        10.00  \n",
       "543        306.7         1.19  \n",
       "580         59.7        10.00  \n",
       "618         82.2        10.00  \n",
       "...          ...          ...  \n",
       "93452       49.7         1.00  \n",
       "93501      155.3        10.00  \n",
       "93625       71.3        10.00  \n",
       "94005      115.1        10.00  \n",
       "94134      230.3        10.00  \n",
       "\n",
       "[864 rows x 18 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from src.auxiliary_functions.VIBRANT_filtering_functions import filter_VIBRANT_annotations_by_mcp\n",
    "# Filter the VIBRANT annotations DataFrame based on MCP-related terms\n",
    "filtered_VIBRANT_annotations = filter_VIBRANT_annotations_by_mcp(VIBRANT_annotations,mcp_terms=mcp_terms_by_usr, false_terms=false_terms_by_usr)\n",
    "filtered_VIBRANT_annotations\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>scaffold</th>\n",
       "      <th>type</th>\n",
       "      <th>Quality</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5711</th>\n",
       "      <td>DTR_889600</td>\n",
       "      <td>lysogenic</td>\n",
       "      <td>complete circular</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5712</th>\n",
       "      <td>DTR_263482</td>\n",
       "      <td>lysogenic</td>\n",
       "      <td>complete circular</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5713</th>\n",
       "      <td>DTR_484729</td>\n",
       "      <td>lysogenic</td>\n",
       "      <td>complete circular</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5714</th>\n",
       "      <td>DTR_767073</td>\n",
       "      <td>lysogenic</td>\n",
       "      <td>complete circular</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5715</th>\n",
       "      <td>DTR_408256</td>\n",
       "      <td>lysogenic</td>\n",
       "      <td>complete circular</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11400</th>\n",
       "      <td>DTR_272404</td>\n",
       "      <td>lytic</td>\n",
       "      <td>complete circular</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11401</th>\n",
       "      <td>DTR_123191</td>\n",
       "      <td>lytic</td>\n",
       "      <td>complete circular</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11402</th>\n",
       "      <td>DTR_630980</td>\n",
       "      <td>lytic</td>\n",
       "      <td>complete circular</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11403</th>\n",
       "      <td>DTR_614252</td>\n",
       "      <td>lytic</td>\n",
       "      <td>complete circular</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11404</th>\n",
       "      <td>DTR_696999</td>\n",
       "      <td>lytic</td>\n",
       "      <td>complete circular</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5694 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         scaffold       type            Quality\n",
       "5711   DTR_889600  lysogenic  complete circular\n",
       "5712   DTR_263482  lysogenic  complete circular\n",
       "5713   DTR_484729  lysogenic  complete circular\n",
       "5714   DTR_767073  lysogenic  complete circular\n",
       "5715   DTR_408256  lysogenic  complete circular\n",
       "...           ...        ...                ...\n",
       "11400  DTR_272404      lytic  complete circular\n",
       "11401  DTR_123191      lytic  complete circular\n",
       "11402  DTR_630980      lytic  complete circular\n",
       "11403  DTR_614252      lytic  complete circular\n",
       "11404  DTR_696999      lytic  complete circular\n",
       "\n",
       "[5694 rows x 3 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from src.auxiliary_functions.VIBRANT_filtering_functions import filter_VIBRANT_genome_quality\n",
    "# Filter the VIBRANT genome quality DataFrame based on the completeness of the genomes\n",
    "filtered_VIBRANT_genome_quality = filter_VIBRANT_genome_quality(VIBRANT_genome_quality, min_quality_by_usr)\n",
    "filtered_VIBRANT_genome_quality"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome_id</th>\n",
       "      <th>data_source</th>\n",
       "      <th>contig_length</th>\n",
       "      <th>dtr_length</th>\n",
       "      <th>genes_viral</th>\n",
       "      <th>genes_plasmid</th>\n",
       "      <th>genes_microbial</th>\n",
       "      <th>genes_unknown</th>\n",
       "      <th>biome1</th>\n",
       "      <th>biome2</th>\n",
       "      <th>biome3</th>\n",
       "      <th>gca_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DTR_003403</td>\n",
       "      <td>IMG/M</td>\n",
       "      <td>4496</td>\n",
       "      <td>91</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>Engineered</td>\n",
       "      <td>Wastewater</td>\n",
       "      <td>Nutrient removal</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DTR_003989</td>\n",
       "      <td>IMG/M</td>\n",
       "      <td>4748</td>\n",
       "      <td>50</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>Host-Associated</td>\n",
       "      <td>Mammals</td>\n",
       "      <td>Digestive system</td>\n",
       "      <td>GCA_001503475.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>DTR_005113</td>\n",
       "      <td>IMG/M</td>\n",
       "      <td>4691</td>\n",
       "      <td>75</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>Intertidal zone</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>DTR_005266</td>\n",
       "      <td>IMG/M</td>\n",
       "      <td>3026</td>\n",
       "      <td>98</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>Host-Associated</td>\n",
       "      <td>Invertebrates</td>\n",
       "      <td>Digestive system</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>DTR_006007</td>\n",
       "      <td>IMG/M</td>\n",
       "      <td>4651</td>\n",
       "      <td>79</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>Neritic zone</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7629</th>\n",
       "      <td>DTR_893354</td>\n",
       "      <td>GOV2</td>\n",
       "      <td>4751</td>\n",
       "      <td>55</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7630</th>\n",
       "      <td>DTR_893395</td>\n",
       "      <td>GOV2</td>\n",
       "      <td>4777</td>\n",
       "      <td>50</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7634</th>\n",
       "      <td>DTR_893440</td>\n",
       "      <td>GOV2</td>\n",
       "      <td>3416</td>\n",
       "      <td>55</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCA_003651745.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7641</th>\n",
       "      <td>DTR_893610</td>\n",
       "      <td>GOV2</td>\n",
       "      <td>4726</td>\n",
       "      <td>55</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7647</th>\n",
       "      <td>DTR_893796</td>\n",
       "      <td>GOV2</td>\n",
       "      <td>3216</td>\n",
       "      <td>55</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Marine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCA_004201515.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1495 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       genome_id data_source  contig_length  dtr_length  genes_viral  \\\n",
       "1     DTR_003403       IMG/M           4496          91            2   \n",
       "4     DTR_003989       IMG/M           4748          50            3   \n",
       "11    DTR_005113       IMG/M           4691          75            1   \n",
       "12    DTR_005266       IMG/M           3026          98            1   \n",
       "19    DTR_006007       IMG/M           4651          79            1   \n",
       "...          ...         ...            ...         ...          ...   \n",
       "7629  DTR_893354        GOV2           4751          55            1   \n",
       "7630  DTR_893395        GOV2           4777          50            1   \n",
       "7634  DTR_893440        GOV2           3416          55            1   \n",
       "7641  DTR_893610        GOV2           4726          55            1   \n",
       "7647  DTR_893796        GOV2           3216          55            1   \n",
       "\n",
       "      genes_plasmid  genes_microbial  genes_unknown           biome1  \\\n",
       "1                 0                0              4       Engineered   \n",
       "4                 0                0              3  Host-Associated   \n",
       "11                0                0              6          Aquatic   \n",
       "12                0                0              2  Host-Associated   \n",
       "19                0                0              7          Aquatic   \n",
       "...             ...              ...            ...              ...   \n",
       "7629              0                0              8          Aquatic   \n",
       "7630              0                0              7          Aquatic   \n",
       "7634              0                0              7          Aquatic   \n",
       "7641              0                0             11          Aquatic   \n",
       "7647              0                0              5          Aquatic   \n",
       "\n",
       "             biome2            biome3         gca_name  \n",
       "1        Wastewater  Nutrient removal              NaN  \n",
       "4           Mammals  Digestive system  GCA_001503475.1  \n",
       "11           Marine   Intertidal zone              NaN  \n",
       "12    Invertebrates  Digestive system              NaN  \n",
       "19           Marine      Neritic zone              NaN  \n",
       "...             ...               ...              ...  \n",
       "7629         Marine               NaN              NaN  \n",
       "7630         Marine               NaN              NaN  \n",
       "7634         Marine               NaN  GCA_003651745.1  \n",
       "7641         Marine               NaN              NaN  \n",
       "7647         Marine               NaN  GCA_004201515.1  \n",
       "\n",
       "[1495 rows x 12 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from src.auxiliary_functions.VIBRANT_filtering_functions import filter_scaffolds_by_contig_length\n",
    "# Filter the JGI global metadata DataFrame based on the length of the contigs\n",
    "filtered_JGI_global_metadata = filter_scaffolds_by_contig_length(JGI_global_metadata, min_contig_length=min_contig_length_by_usr, max_contig_length=max_contig_length_by_usr)\n",
    "filtered_JGI_global_metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_and_write_MCP_sequences(filtered_VIBRANT_annotations, VIBRANT_phages_proteins, output_path):\n",
    "    # Read the column \"protein\" from filtered_VIBRANT_annotations and create a list of MCP IDs\n",
    "    filtered_VIBRANT_annotations_MCPs_IDS = filtered_VIBRANT_annotations['protein'].tolist()\n",
    "    \n",
    "    # Filter the fasta sequences from VIBRANT_phages_proteins by retaining only relevant IDs\n",
    "    filtered_sequences = [seq_record for seq_record in VIBRANT_phages_proteins if seq_record.id in filtered_VIBRANT_annotations_MCPs_IDS]\n",
    "    \n",
    "    # Write the filtered sequences to a fasta file at the specified output path\n",
    "    SeqIO.write(filtered_sequences, output_path, \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "vibrant_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
